Fix #20893: Prevent fast-click events for newsletter subscription

tiagombrbrito opened the pull request:
## Overview

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #20893 .
2. This PR does the following: Disables the "Subscribe" button while processing a subscription, re-enabling the "Subscribe" button after request response is handled. Adds a local storage of submitted emails for newsletter subscription, to prevent fast-click events and multiple submissions of already submitted email addresses. Adds ngx-toastr/toastr CSS to root's styles for lightweight-oppia-root's toastr style importing.
3. (For bug-fixing PRs only) The original bug occurred because: Splash page's toastr module CSS isn't being bundled by webpack for lightweight-oppia-root. No debouncing mechanism was added to subscribeToMailingList and "Subscribe" button, allowing fast-click events for newsletter subscription.

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

Correct CSS for info toasts on Splash page with disabling button and blocking requests of already submitted emails

[issue20893v2-chrome-withoutmodal.webm](https://github.com/user-attachments/assets/11af8661-7761-4918-875d-091a98c99e21)

Disabling mechanisms and blocking requests of already submitted emails with modal production

[issue20893v2-chrome-withmodal.webm](https://github.com/user-attachments/assets/a04d4539-5a48-414f-a780-a6365d34b59c)

UI width transition

[issue20893v2-chrome-UItransition.webm](https://github.com/user-attachments/assets/ae091f35-4547-47d3-ac80-bc1a6b0d6389)

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

[issue20893v2-chrome-throttled.webm](https://github.com/user-attachments/assets/d8844a08-a51a-4f1e-a02f-3283e84ac0ce)

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->
~~No proof of changes for mobile or Arabic language needed because this PR doesn't change the UI, only fixes incorrect toastr displays and adds debouncing mechanism to newsletter subscription.~~

#### Proof of changes on mobile phone
Correct CSS for info toasts on Splash page with disabling button and blocking requests of already submitted emails

[issue20893v2-chrome-mobile-withoutmodal.webm](https://github.com/user-attachments/assets/de071d7b-98d7-443d-8611-36a7bb35d19e)

Disabling mechanisms and blocking requests of already submitted emails with modal production

[issue20893v2-chrome-mobile-withmodal.webm](https://github.com/user-attachments/assets/13dbbfa4-7a6f-4a1b-b4bf-f2d7d25b0143)

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

[issue20893v2-chome-rtl-withmodal.webm](https://github.com/user-attachments/assets/8e9a5bf5-2822-44ac-8258-086500267336)

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

tiagombrbrito committed:
Fix fast-click events for newsletter subscription

- Fix splash page ngx-toastr CSS import
- Add boolean subscriptionProcessing to OppiaFooterComponent
- Add function disableNewsletterSubscription to OppiaFooterComponent
- Update function subscribeToMailingList in OppiaFooterComponent
- Add front-end tests for disableNewsletterSubscription
- Update front-end tests for subscribeToMailingList

tiagombrbrito commented:
@mon4our PTAL

mon4our commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -68,34 +70,48 @@ export class OppiaFooterComponent {
-        if (status) {
-          this.alertsService.addInfoMessage('Done!', 1000);
-          this.ngbModal.open(ThanksForSubscribingModalComponent, {
-            backdrop: 'static',
-            size: 'xl',
-          });
-        } else {
+        setTimeout(() => {
+          if (status) {
+            this.alertsService.addInfoMessage('Done!', 1000);
+            this.ngbModal.open(ThanksForSubscribingModalComponent, {
+              backdrop: 'static',
+              size: 'xl',
+            });
+          } else {
+            this.alertsService.addInfoMessage(
+              AppConstants.MAILING_LIST_UNEXPECTED_ERROR_MESSAGE,
+              10000
+            );
+          }
+          this.subscriptionProcessing = false;
+        }, 500);
+      })
+      .catch(errorResponse => {
+        setTimeout(() => {
           this.alertsService.addInfoMessage(
             AppConstants.MAILING_LIST_UNEXPECTED_ERROR_MESSAGE,
             10000
           );
-        }
-      })
-      .catch(errorResponse => {
-        this.alertsService.addInfoMessage(
-          AppConstants.MAILING_LIST_UNEXPECTED_ERROR_MESSAGE,
-          10000
-        );
+          this.subscriptionProcessing = false;
+        }, 500);
       });
   }
```
Instead of doing this can't we just show the success or failure message as soon as the response from the subscription is received (like we were doing previously)? We should avoid using a setTimeout to make the user experience as smooth as possible. If a response is received before the setTimeout expires, the user will have to wait the remaining time before seeing the appropriate message which is not the best ux.

tiagombrbrito replied:
The idea was to enforce a delay in the generated response to prevent rapid submissions of email subscriptions. While the modal acts as a block to rapid submissions, the scenarios of failure/rejection of subscription are rather quick, allowing for such behavior. So I implemented a delay that wouldn't harm the UX severely but would prevent rapid submissions. That said, if you feel that the UX shouldn't be disturbed over this, I can remove the delay.

seanlip replied:
@tiagombrbrito I think that these hardcoded delays can be flaky. If you want to prevent repeated submissions, might I suggest a different approach -- store the list of emails that have been submitted and that resulted in a "success" response as a local variable. Then, if the user tries to submit that email again, they see an error saying "You have already signed up with this email address." That only lasts as long as they are still on the page.

What do you think about this approach?

tiagombrbrito replied:
This seems like a good approach to avoid backend flooding by rejected email requests. Should the message be displayed in a toastr or a new modal?

seanlip replied:
I think small red text under the input field should be fine (or something along those lines, not sure how well red displays on dark green).

tiagombrbrito replied:
Red doesn't look good with this green. I also tried the footer's link color and another lighter orange to keep up with the color scheme, but didn't have enough contrast with the background. The only two options I've come across are either keeping with the white (which doesn't produce a warning effect but is cohesive with the color scheme) or a yellow (which produces more attention but isn't so cohesive with the color scheme). Let me know your thoughts.

![Screenshot from 2025-05-28 15-45-03](https://github.com/user-attachments/assets/ecbcbb0d-d5d2-4305-a73d-7dc59245abd9)

![Screenshot from 2025-05-28 15-48-51](https://github.com/user-attachments/assets/4cb34649-6a0e-4378-8cb6-9996fd0de09c)

seanlip replied:
I think let's go with the second one for now, we can make changes in future if needed. Thanks for exploring the options!

tiagombrbrito replied:
Hi, I'm getting blocked from pushing my changes because of a back-end test failing. I'm aware there is an open issue for this #20802 . Is there a temporary fix available for it?

![Screenshot from 2025-06-04 01-03-02](https://github.com/user-attachments/assets/fc912c83-336f-42e5-99c3-c517125d91ed)

tiagombrbrito replied:
Done! Updated the PR description to showcase the requested changes.

mon4our replied:
Hey @tiagombrbrito just wanted to check, how were you able to get past the issue you described [here](https://github.com/oppia/oppia/pull/22664#discussion_r2125155464)?

tiagombrbrito replied:
Hi @mon4our , i was able to bypass the back-end test failure by manually altering the system's timezone to an UTC timezone.
![Screenshot from 2025-06-12 15-17-21](https://github.com/user-attachments/assets/58f9a78c-5af1-4a53-9ee7-3fdfc96f303a)

tiagombrbrito committed:
Merge updates from upstream develop into issue20893

tiagombrbrito committed:
Address review comments

- Add i18n variable for newsletter warning
- Add ngModelChange for newsletter input
- Add CSS for oppia-footer-newsletter-warning
- Remove delay from subscribeToMailingList
- Add boolean emailDuplicated to OppiaFooterComponent
- Add set emailsSubscribed to OppiaFooterComponent
- Add function isAlreadySubscribed to OppiaFooterComponent
- Add function clearNewsletterWarning to OppiaFooterComponent
- Update front-end tests for subscribeToMailingList
- Add front-end tests for isAlreadySubscribed
- Add front-end test for clearNewsletterWarning

tiagombrbrito committed:
Merge updates from branch develop into issue20893

tiagombrbrito committed:
Fix add email to set

tiagombrbrito commented:
@mon4our PTAL

mon4our commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -68,35 +72,58 @@ export class OppiaFooterComponent {
+      this.emailDuplicated = true;
+    } else {
+      this.subscriptionProcessing = true;
+      this.emailsSubscribed.add(this.emailAddress ?? '');
```
Why is the condition:
```typescript
(this.emailAddress ?? '')
``` 
Not just:
```typescript
(this.emailAddress)
``` 
I feel the `emailsSubscribed` Set should only contain emails. Maybe try showing an error message that the email address entered is null/undefined.

tiagombrbrito replied:
Typescript enforces this due to emailAddress being typed as null | string. When subscribeToMailingList() is called by clicking the "Subscribe" button, validateEmailAddress() ensures that emailAddress has a valid email; otherwise, the "Subscribe" button would be disabled. So, I don't think there is a situation that warrants an invalid email warning. That being said, I can recast emailAddress as a string as done in line 72.

tiagombrbrito replied:
Done. I've recast emailAddress as a string, and given the context I shared before, I hope that's enough to address your concerns.

mon4our commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -68,35 +72,58 @@ export class OppiaFooterComponent {
+  }
+
+  isAlreadySubscribed(email: string | null): boolean {
+    return this.emailsSubscribed.has(email ?? '');
```
Ditto below. Keep only valid emails that have successfully subscribed in this Set.

tiagombrbrito replied:
Done.

mon4our reviewed (changes requested):
The warning message you implemented looks good. Just wanted to raise a concern regarding the `emailsSubscribed` Set. Thanks!

tiagombrbrito committed:
Merge updates from branch develop into issue20893

tiagombrbrito committed:
Merge updates from develop into issue20893

tiagombrbrito committed:
Fix emailAddress type handling

mon4our reviewed (approved):
LGTM!

mon4our committed:
Merge branch 'develop' into issue20893

tiagombrbrito commented:
Full-stack tests / E2E (wipeout) and Full-stack tests / Acceptance (exploration-editor/submit-review-and-respond-to-feedback) failed, already left a comment in the following issues: 
- E2E: https://github.com/oppia/oppia/issues/16640#issuecomment-2992391309
- Acceptance: https://github.com/oppia/oppia/issues/22822#issuecomment-2992386944

Could a maintainer restart the failed tests? 
Thank you
